[MRG+1] Raise error when SparseSeries is passed into classification metrics

nielsenmarkus11 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes #7352 
#### What does this implement/fix? Explain your changes.

This change raises an error when the type is a pandas SparseSeries of either the `y_true` or `y_score` input.
#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

mniels17 committed:
Raise error when SparseSeries is passed into roc_curve

nielsenmarkus11 committed:
Changed "y_true" in second if block to "y_score"

maniteja123 commented on sklearn/metrics/ranking.py:
```diff
@@ -22,6 +22,7 @@
 import warnings
 import numpy as np
 from scipy.sparse import csr_matrix
+import pandas
```
Hi, thanks for the PR. But AFAIR, it is a convention in general to import pandas and also put it in a try catch block ( like [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_grid_search.py#L491) ). I suppose this is necessary since Pandas is not a compulsory requirement for installing scikit-learn and this import would raise an error in case Pandas isn't installed. Other than that, LGTM.

amueller replied:
actually, you can not import pandas AT ALL. What you refer to is a test, and you can't really do that same thing here. You can check the class name, though, if you like.

maniteja123 replied:
@amueller sorry for my ignorance. Will keep it in mind from next time.

amueller replied:
@maniteja123 don't sweat it, any help is appreciated :)

nielsenmarkus11 replied:
Okay. I will be removing this line.  Thanks.

jnothman commented:
This shouldn't just be in `roc_*` but should fix all metrics by putting in `type_of_target`. Needs tests.

nielsenmarkus11 commented:
@jnothman, just for clarity, since `roc_*` doesn't check using the `type_of_target` would you still recommend adding the call to the function within `roc_*`?

Thanks

mniels17 committed:
Remove code to import pandas and add sparseseries check to 'type_of_target' function. Finally, add 'type_of_target' call to _binary_clf_curve

nielsenmarkus11 committed:
Merge pull request #1 from scikit-learn/master

Update from Original

nielsenmarkus11 committed:
Merge pull request #2 from nielsenmarkus11/master

Update from Original

mniels17 committed:
Merge branch 'sparse' of https://github.com/nielsenmarkus11/scikit-learn into sparse

mniels17 committed:
Remove pandas import and old comparison in roc_curve.

nielsenmarkus11 commented:
I've removed the `import pandas` code and incorporated a check within the `type_of_target` code.  Tested both in Python 3.5.2 and 2.7.12.  Are there any additional tests that I need to do?

amueller commented:
you should add a test. I'm not sure if we have a mock for a sparse series but there are several pandas mocks that you could use.

nielsenmarkus11 commented:
Just to make sure I understand correctly... I should add a new test to this code in [sklean/utils/tests/test_multiclass.py](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/tests/test_multiclass.py), correct?
```
def test_type_of_target():
    for group, group_examples in iteritems(EXAMPLES):
        for example in group_examples:
            assert_equal(type_of_target(example), group,
                         msg=('type_of_target(%r) should be %r, got %r'
                              % (example, group, type_of_target(example))))

    for example in NON_ARRAY_LIKE_EXAMPLES:
        msg_regex = 'Expected array-like \(array or non-string sequence\).*'
        assert_raises_regex(ValueError, msg_regex, type_of_target, example)

    for example in MULTILABEL_SEQUENCES:
        msg = ('You appear to be using a legacy multi-label data '
               'representation. Sequence of sequences are no longer supported;'
               ' use a binary array or sparse matrix instead.')
        assert_raises_regex(ValueError, msg, type_of_target, example)
```

amueller commented:
yeah that sounds like a reasonable place

mniels17 committed:
Add test for 'type_of_target' function

nielsenmarkus11 commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -289,6 +289,14 @@ def test_type_of_target():
+        from pandas import SparseSeries
+    except ImportError:
+        pass
+    y = SparseSeries([1,0,0,1,0])
```
So I'm seeing the error in the automatic checks that states: "UnboundLocalError: local variable 'SparseSeries' referenced before assignment"  Do I need to put all of the test code within the `try` block?

nielsenmarkus11 commented on sklearn/metrics/ranking.py:
```diff
@@ -294,6 +294,11 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
     """
+    # Check to make sure y_true is valid
+    y_type = type_of_target(y_true)
+    if y_type != "binary":
```
Also, in some of the other testing I'm getting the error: `ValueError: multiclass format is not supported`. I was under the impression that the `_binary_clf_curve` required 'binary' data.  Should it also be allowed to accept 'multiclass' data?

mniels17 committed:
Add white space after commas

nielsenmarkus11 commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -289,6 +289,14 @@ def test_type_of_target():
+        from pandas import SparseSeries
+    except ImportError:
+        pass
+    y = SparseSeries([1, 0, 0, 1, 0])
```
So I'm seeing the error in the automatic checks that states: `UnboundLocalError: local variable 'SparseSeries' referenced before assignment` Do I need to put all of the test code within the try block?

mniels17 committed:
Correct other white space issues

mniels17 committed:
Move type_of_target test into try clause, remove test_precision_recall_curve_pos_label since as multiclass it doesn't make sense

mniels17 committed:
Add test_precision_recall_curve_pos_label back in and also add test_binary_clf_curve to test new logic in _binary_clf_curve function

mniels17 committed:
Correct syntax and formatting.

mniels17 committed:
Remove trailing white space

mniels17 committed:
Correct validation logic

nielsenmarkus11 commented:
Okay. I think it should be good now. I understand the supposed issue with `test_precision_recall_curve_pos_label` and put this back in.  Perhaps the documentation can be updated to include the `pos_label` exception for the `y_true` input.  I've added the test back in as well as a `test_binary_clf_curve` function.

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -290,6 +290,13 @@ def test_type_of_target():
+        msg = "y cannot be class 'SparseSeries'."
+        assert_raises_regex(ValueError, msg, type_of_target, y)
+    except ImportError:
+        pass
```
Please wrap only the import statement and use `raise SkipTest("Pandas not found")` as elsewhere

jnothman reviewed (approved):
Otherwise, this LGTM.

nielsenmarkus11 committed:
Update test_multiclass.py per @jnothman 's request.

nielsenmarkus11 commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -290,6 +290,14 @@ def test_type_of_target():
+    y = SparseSeries([1, 0, 0, 1, 0])
+    msg = "y cannot be class 'SparseSeries'."
+    assert_raises_regex(ValueError, msg, type_of_target, y)
 
```
Per @jnothman 's request, I've only wrapped the import and added `raise SkipTest("Pandas not found")` otherwise.

nielsenmarkus11 committed:
Import SkipTest function.

mniels17 committed:
Merge remote-tracking branch 'upstream/master' into sparse

nielsenmarkus11 committed:
Remove extra white space from line 303

jnothman reviewed (approved):
LGTM

GaelVaroquaux commented on sklearn/metrics/ranking.py:
```diff
@@ -294,6 +294,12 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
+    y_type = type_of_target(y_true)
+    if not (y_type == "binary" or
+            (y_type == "multiclass" and pos_label is not None)):
+        raise ValueError("{0} format is not supported".format(y_type))
```
It's a nitpick, but it would help the user to give a different error message if y_type == "multiclass" and pos_label is None.

Beside, I am surprised, but it is really the case that in multiclass settings we require the pos_label not to be specified? I would have though the opposite. Is there an error in the condition above, or in my assumptions on our code?

GaelVaroquaux replied:
As @jnothman pointed out: the code is correct, I was confused by the double negation.

Still, a different error message would help.

lesteve replied:
> Still, a different error message would help.

Agreed. The error message template I generally try to follow is something like:
```
Allowed values for parameter_name are ['value1', 'value2', 'value3']. Instead you provided 'parameter_name={parameter_value}'
```

nielsenmarkus11 replied:
@lesteve, my choice of error message was copy pasted from other portions of this code.  I chose the language to be consistent with the other instances of similar errors in `ranking.py`.

lesteve commented:
Hmmm I am a bit confused on this one. I commented on the issue, see https://github.com/scikit-learn/scikit-learn/issues/7352#issuecomment-305472045.

jnothman commented:
Gael i think the confusion is yours in reading a double negation

On 1 Jun 2017 9:59 pm, "Loïc Estève" <notifications@github.com> wrote:

> Hmmm I am a bit confused on this one. I commented on the issue, see #7352
> (comment)
> <https://github.com/scikit-learn/scikit-learn/issues/7352#issuecomment-305472045>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7373#issuecomment-305472552>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62XaeqMndbeD52eOJWf6AAP9osh8ks5r_qejgaJpZM4J4jvQ>
> .
>

nielsenmarkus11 commented:
Okay... everything is good to go.

lesteve commented on sklearn/utils/multiclass.py:
```diff
@@ -234,6 +234,10 @@ def type_of_target(y):
         raise ValueError('Expected array-like (array or non-string sequence), '
                          'got %r' % y)
 
+    sparseseries = (y.__class__.__name__ == 'SparseSeries')
```
Just testing the name of the class is a bit dodgy, I think it would be better to use an isinstance.

nielsenmarkus11 replied:
Change this to `isinstance(y,pd.sparse.series.SparseSeries)`

nielsenmarkus11 replied:
So, I've gone back to using the name of the class, per prior comments from @amueller on commit [d21c7e38674388f97e146aef67f42bef2fe5d2d2](https://github.com/scikit-learn/scikit-learn/pull/7373/commits/d21c7e38674388f97e146aef67f42bef2fe5d2d2).  Per his comments `pandas` should not be imported at all except in test.

jnothman commented:
You could be more specific, if that's a concern, by checking the type's
module name starts with 'pandas.'. you could also avoid __class__ with
type() etc. You could check the object's mro to see if it is a subclass of
SparseSeries.
, as I've said, I think this is fine

On 2 Jun 2017 7:31 am, "nielsenmarkus11" <notifications@github.com> wrote:

> *@nielsenmarkus11* commented on this pull request.
> ------------------------------
>
> In sklearn/utils/multiclass.py
> <https://github.com/scikit-learn/scikit-learn/pull/7373#discussion_r119734071>
> :
>
> > @@ -234,6 +234,10 @@ def type_of_target(y):
>          raise ValueError('Expected array-like (array or non-string sequence), '
>                           'got %r' % y)
>
> +    sparseseries = (y.__class__.__name__ == 'SparseSeries')
>
> So, I've gone back to using the name of the class, per prior comments from
> @amueller <https://github.com/amueller> on commit
> d21c7e38674388f97e146aef67f42bef2fe5d2d2
> <https://github.com/scikit-learn/scikit-learn/pull/7373/commits/d21c7e38674388f97e146aef67f42bef2fe5d2d2>.
> Per his comments pandas should not be imported at all except in test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7373#discussion_r119734071>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67_7fTexHGOhdpageYjs3sFNq_shks5r_y2ZgaJpZM4J4jvQ>
> .
>

nielsenmarkus11 committed:
Merge branch 'master' into sparse

jnothman commented:
Another review here?

nielsenmarkus11 commented:
Ping @amueller

GaelVaroquaux commented:
LGTM. Merging given @jnothman 's +1

GaelVaroquaux commented:
Thanks!

nielsenmarkus11 commented:
Thank you!
